// VoiceIt/index.js
var request = require("request");
var crypto = require('crypto');
var fs = require('fs');

function getSHA256(passwd)
{
	var hash = crypto.createHash('sha256').update(passwd).digest('hex');
	return hash;
}

function checkUndefined(param)
{
	if (param == null){
		return "";
	}
	else{
		return param;
	}
}

module.exports =
{
	developerId: '',
	initialize: function(devId)
	{
	  this.developerId = devId;
	  console.log("Variables Intitialized");
	},
	getUser: function(credentials)
	{
		var password = getSHA256(credentials.password);
		request({
		    headers: {
						'PlatformID': '5',
		        'Content-Length': '0',
		        'Content-Type': 'application/json',
				'UserId':credentials.userId,
				'VsitPassword':password,
				'VsitDeveloperId':this.developerId
		    },
		    uri: 'https://siv.voiceprintportal.com/sivservice/api/users',
		    body: '',
		    method: 'GET'}
		  , function (error, response, body) {
		      if (credentials.callback && typeof(credentials.callback) === "function") {
		          credentials.callback(body);
		      }
		});
	},
	deleteUser: function(credentials)
	{
		var password = getSHA256(credentials.password);
		request({
		    headers: {
						'PlatformID': '5',
		        'Content-Length': '0',
		        'Content-Type': 'application/json',
				'UserId':credentials.userId,
				'VsitPassword':password,
				'VsitDeveloperId':this.developerId
		    },
		    uri: 'https://siv.voiceprintportal.com/sivservice/api/users',
		    body: '',
		    method: 'DELETE'}
		  , function (error, response, body) {
		      if (credentials.callback && typeof(credentials.callback) === "function") {
		          credentials.callback(body);
		      }
		});
	},
	createUser: function(credentials)
	{
		var password = getSHA256(credentials.password);
		request({
		    headers: {
						'PlatformID': '5',
		        'Content-Length': '0',
		        'Content-Type': 'application/json',
				'UserId':credentials.userId,
				'VsitPassword':password,
				'VsitDeveloperId':this.developerId
		    },
		    uri: 'https://siv.voiceprintportal.com/sivservice/api/users',
		    body: '',
		    method: 'POST'}
		  , function (error, response, body) {
		      if (credentials.callback && typeof(credentials.callback) === "function") {
		          credentials.callback(body);
		      }
		});
	},
	createEnrollment: function(credentials)
	{
		var password = getSHA256(credentials.password);
			data = fs.readFileSync(credentials.pathToEnrollmentWav);
			request({
			headers: {
			'PlatformID': '5',
			'Content-Type': 'audio/wav',
			'UserId':credentials.userId,
			'VsitPassword':password,
			'VsitDeveloperId':this.developerId,
			'ContentLanguage': checkUndefined(credentials.contentLanguage),
			},
			uri: 'https://siv.voiceprintportal.com/sivservice/api/enrollments',
			body: data,
			method: 'POST'}
			, function (error, response, body) {
			if (credentials.callback && typeof(credentials.callback) === "function") {
			  credentials.callback(body);
			}
			});
	},
	createEnrollmentByWavURL: function(credentials)
	{
		var password = getSHA256(credentials.password);
			request({
			headers: {
			'PlatformID': '5',
			'Content-Type': 'audio/wav',
			'UserId':credentials.userId,
			'VsitPassword':password,
			'VsitDeveloperId':this.developerId,
			'VsitwavURL': credentials.urlToEnrollmentWav,
			'ContentLanguage': checkUndefined(credentials.contentLanguage),
			},
			uri: 'https://siv.voiceprintportal.com/sivservice/api/enrollments/bywavurl',
			body: '',
			method: 'POST'}
			, function (error, response, body) {
			if (credentials.callback && typeof(credentials.callback) === "function") {
			  credentials.callback(body);
			}
			});
	},

	getEnrollments: function(credentials)
	{
		var password = getSHA256(credentials.password);
			request({
			headers: {
			'PlatformID': '5',
			'Content-Type':'application/json; charset=UTF-8',
			'Accept':'application/json',
			'Encoding':'utf8',
			'UserId':credentials.userId,
			'VsitPassword':password,
			'VsitDeveloperId':this.developerId,
			},
			uri: 'https://siv.voiceprintportal.com/sivservice/api/enrollments/',
			body: '',
			method: 'GET'}
			, function (error, response, body) {
			if (credentials.callback && typeof(credentials.callback) === "function") {

				credentials.callback(response.body);
			}
			});
	},
	deleteEnrollment: function(credentials)
	{
		var password = getSHA256(credentials.password);
			request({
			headers: {
			'PlatformID': '5',
			'Content-Type': 'application/json',
			'UserId':credentials.userId,
			'VsitPassword':password,
			'VsitDeveloperId':this.developerId,
			},
			uri: 'https://siv.voiceprintportal.com/sivservice/api/enrollments/'+credentials.enrollmentId,
			body: '',
			method: 'DELETE'}
			, function (error, response, body) {
			if (credentials.callback && typeof(credentials.callback) === "function") {
			  credentials.callback(body);
			}
			});
	},
	authentication: function(credentials)
	{
		var password = getSHA256(credentials.password);
			data = fs.readFileSync(credentials.pathToAuthenticationWav);
			request({
			headers: {
			'PlatformID': '5',
			'Content-Type': 'audio/wav',
			'UserId':credentials.userId,
			'VsitPassword':password,
			'VsitDeveloperId':this.developerId,
			'ContentLanguage':checkUndefined(credentials.contentLanguage),
			},
			uri: 'https://siv.voiceprintportal.com/sivservice/api/authentications',
			body: data,
			method: 'POST'}
			, function (error, response, body) {
			if (credentials.callback && typeof(credentials.callback) === "function") {
			  credentials.callback(body);
			}
			});
	},
	authenticationByWavURL: function(credentials)
	{
		var password = getSHA256(credentials.password);
			request({
			headers: {
			'PlatformID': '5',
			'Content-Type': 'audio/wav',
			'UserId':credentials.userId,
			'VsitPassword':password,
			'VsitDeveloperId':this.developerId,
			'VsitwavURL':credentials.urlToAuthenticationWav,
			'ContentLanguage':checkUndefined(credentials.contentLanguage),
			},
			uri: 'https://siv.voiceprintportal.com/sivservice/api/authentications/bywavurl',
			body: '',
			method: 'POST'}
			, function (error, response, body) {
			if (credentials.callback && typeof(credentials.callback) === "function") {
			  credentials.callback(body);
			}
			});
	},
	getEnrollmentsCount: function(credentials)
	{
		var password = getSHA256(credentials.password);
			request({
			headers: {
			'PlatformID': '5',
			'Content-Type':'application/json; charset=UTF-8',
			'Accept':'application/json',
			'Encoding':'utf8',
			'UserId':credentials.userId,
			'VsitPassword':password,
			'VsitDeveloperId':this.developerId,
			'VppText':credentials.phrase
			},
			uri: 'https://siv.voiceprintportal.com/sivservice/api/enrollments/count',
			body: '',
			method: 'GET'}
			, function (error, response, body) {
			if (credentials.callback && typeof(credentials.callback) === "function") {

				credentials.callback(response.body);
			}
			});
	}
};
